### ISSUE TITLE
Pseudo-tree-shaking

### ISSUE THREAD

**Issue by jgaskins** [opal/opal#1734]
I had a look at the unused code in the Opal corelib:

<img width="1265" alt="Unused code report" src="https://user-images.githubusercontent.com/108205/34594045-b5c992da-f19c-11e7-976f-4ace19f297f8.png">

It's a lot.

This was for a Clearwater app on 0.10, but 0.11 is probably along the same lines. I was wondering if it might make sense to do a little bit of tree-shaking to filter out methods with names that are never called on any class. It wouldn't be perfect because of duck typing (for example, `MyClass#call` would make it into the build even if it were never called because `Proc#call` is used everywhere), but if it's possible to filter out method definitions and shrink the payload, total bundle size could probably be cut in half in a significant number of apps.

I'm not sure how this could work, though. I was thinking something like `compiler.method_calls`, but on a larger scale. It also means (and this is probably the hard part) that an _entire application_ would need to be parsed for method calls before any code was generated. I have a feeling that would require significant changes to the compiler — IIRC, the compiler generates the code as it parses, right?

If this can't be done, would a post-processor that grabs the `add_stubs` calls from a bundle and removes all calls to `Opal.defn` for methods that aren't in that set make sense?

**Comment by jgaskins** [comment]
Took a shot at post-processing my concatenated/minified app bundle. Looks like this might take a bit more than just basing it on on `compiler.method_calls` / `Opal.add_stubs` to have much effect. Consider the following code:

```ruby
class Foo
  def bar
    baz
  end

  def baz
  end
end
```

In this example, `baz` is not removed because it's called by `bar`, but if nothing ever calls `bar`, `bar` gets removed, but `baz` is never actually invoked and should be removed.

In a quick attempt at doing this locally on my app bundle, I get: `{:stubs=>1018, :method_defs=>438, :filtered=>27}` — only about 6% of methods that would be filtered out. I don't know how many bytes that is, though, because I haven't parsed functions yet. This is just a quick scan with regexes so far.

I'm assuming the high ratio of stubs to defs is due to metaprogramming, but I'm not certain yet. That means it shouldn't get too many false positives, though, which is good. Should be able to whitelist any false positives that do occur, though, due to dynamic calls via `send` or whatever.

All this is to say that any tree shaking may need to make multiple passes and not be based on `add_stubs`.

**Comment by jgaskins** [comment]
After a little more experimentation, I was able to come up with a [multipass post-processor](https://gist.github.com/jgaskins/b946e89fe0c8e05c68c34050c38f0541) that filters out definitions of methods whose names are not used anywhere statically and was even able to account for a few dynamic places.

For a baseline Clearwater app, I was able to take the post-minification (not gzipped) size from 413KB to 354KB (~14%). Not bad.

One thing I noticed, though, was that `Opal.add_stubs` takes the largest chunk of startup time. When I cranked up Chrome's slow-my-shit-down feature, this is what it looks like:

<img width="424" alt="Before filter" src="https://user-images.githubusercontent.com/108205/34599504-added584-f1c0-11e7-8fc7-11ae37f5dded.png">

So we're looking at a little over 200ms just for `add_stubs` and GC when the app boots. After running it through that filter (which also yanks out `add_stubs` calls), this happened:

<img width="424" alt="After filter" src="https://user-images.githubusercontent.com/108205/34599545-f5654140-f1c0-11e7-9437-06d4330d3c17.png">

So after patching this app post-compile/minification, `add_stubs` takes 30% less time (counting the full stack underneath it because we're eliminating calls to it entirely), initial GC drops by a whopping 75% (going from second place to _fifth_), and total startup performance is improved by about 25% (580ms down from 767ms).

I have no idea what to do with it from here, though, tbh. This is just a proof of concept after a few hours of screwing around.

**Comment by iliabylich** [comment]
@jgaskins We had a similar experiment before: https://github.com/opal/opal/pull/1357 and the conclusion was: it's doable, but only if there's no metaprogramming (like `send` or `eval`). If you have some magic on your code (and most probably you do) it just doesn't make sense. As always, the problem here comes from the nature of the Ruby, it's too dynamic. You can't determine on the compilation stage which methods are used by the code and which aren't.

By the way, you can drop all `Opal.add_stubs` invocations by disabling `method_missing` in the config (`--no-method-missing` for CLI).

I've just realized that we had a similar discussion in https://github.com/opal/opal/issues/1588.

**Comment by jgaskins** [comment]
Yeah, the more I thought about it, I realized that doing it during the compilation phase would be difficult to get right, as mentioned in those other discussions. Running it as a post-compilation task gave me the benefit of seeing the whole app at once after everything was known about it. This was handy because I am more familiar with the compiled code than with the compiler internals.

One advantage of this approach over Closure Compiler and Prepack is that it’s written for Opal compiled code rather than for idiomatic JavaScript code. There are things Opal does that hand-written JS would never do (`Opal.defn(self, “$foo”, ...)` vs `Thing.prototype.foo = ...`), so CC and Prepack wont understand how to optimize it.

> but only if there's no metaprogramming

Yep, I addressed this in my second post: “Should be able to whitelist any false positives that do occur, though, due to dynamic calls via `send` or whatever.” My script allows adding method names to a list to keep from removing them. Classic case of an optimization requiring a bit of hand-tuning of the results when they can’t be generalized, and I tried to make it as painless as possible.

I do a fair bit of metaprogramming in Clearwater apps with things like `button({ onclick: method(:foo) })`, where we need to keep `foo` despite it never being invoked directly. My script accounts for this case, too, as long as the method name is hardcoded. I didn’t try accounting for `send` at all because that’s primarily invoked with a dynamic method name, but I’m sure it could be handled reasonably easily.

> you can drop all `Opal.add_stubs` invocations by disabling `method_missing`

The idea was to filter out `add_stubs` calls that we’re never used rather than remove them entirely. They may be slow, but they’re handy and `Bowser` relies on `method_missing` for elements and events to delegate to their native counterparts.

**Comment by ahmadine** [comment]
@jgaskins I updated your script and my results were 210 kB minified -> 110 kB minified. I may have a different usecase than most here (my code is tiny and I only include parts of corelib) and I had to do some changes for 1.0.0 compatibility, but the results are entirely satisfying. Thank you :)

https://gist.github.com/ahmadine/c6238af75c2a43a5f89032820b8eed01

**Comment by jgaskins** [comment]
Whoa, that’s impressive! Looks like you were able to handle a few more cases, too! Nice work!

**Comment by ahmadine** [comment]
I worked a bit more on the idea, I managed to also implement (in a very hacky way) procedures to clean up unused:

* constants
* classes and modules
* dangling JS functions
* runtime functions

I have yet to test it, but I don't think it really improves the big codebases, but it works very well with especially small, carefully crafter for such an usage, codebases. I want to clean that up and make a small PoC involving a tiny codebase and see it being stripped as much as possible to the barebones.

This kind of brings us to the point, that for Opal 1.2, if we want to work with cleaning up the runtime and dangling JS functions, we would need to work on cleaning up the tree AFTER generating the JS code which would perfectly involve invoking some JS parser. Do note, that it would need to somehow involve the webpack efforts and especially #2018 (which is a great idea, but just interferes a bit with what I'm trying to do here, we would probably need to work with a bigger awareness of webpack parts - and at the current time I have almost no knowledge about webpack internals).

This code includes a JS parser and looks more or less related to our efforts: https://github.com/i10a/minjs This parser probably doesn't work with ES2015 (generator functions), ES2017 (async/await), but we currently target ES5 Strict which should be enough.

**Comment by ahmadine** [comment]
@jgaskins Here's the promised demo:

https://github.com/ahmadine/opal-multishake-demo

**Comment by iliabylich** [comment]
I think it's possible to use AST to find dead constants/methods (if they are defined without using any kind of metaprogramming or raw JS). About a year ago I wrote this script to find unused constants in the Rails app - https://gist.github.com/iliabylich/06c0a67244553742efd06d069fc70cb6, maybe it can be useful for someone. It holds `Module.nesting` under the hood and so it properly computes class names for nested constants (unless you use `module_eval` 😄 ).

It can be extended to also register used methods (and thus, eventually find unused methods).

Probably it can be also combined with a JS parser to handle inline JS and method calls that start with a `$` character.

**Comment by ahmadine** [comment]
@iliabylich I would rather do it entirely on JS level and I will probably do it someday (maybe soon).

I mentioned minjs before, it involves a nice parser, I would probably reuse it from this library, while keeping the Opal logic separate.

This is kind of how it would look like:

https://github.com/i10a/minjs/blob/master/lib/minjs/compressor/compressor.rb

**Comment by ahmadine** [comment]
Regarding the JS parsers available in Ruby: I don't think minjs is the best bet for that. I explored some other gems that I may use for this project:

* https://github.com/tenderlove/recma - used to be called RKelly, renamed just **3 days ago**, but otherwise outdated. It kind of looks like @tenderlove is going to revive this project
* https://github.com/nene/rkelly-remix - an RKelly fork which includes the features I would really like to use (last commit in 2017) - and with this featureset and what I learned with improving the @jgaskins code, implementing an Opal shaker should be quite easy

Other JS parsers:
* https://github.com/nex3/rbnarcissus - really small and looks nice, but severely outdated. apparently RKelly was originally based on that
* https://github.com/wycats/parsejs - quite nice, but outdated
